upstairs/src/volume.rs

```diff
@@ -372,28 +372,32 @@ impl Volume {
                 }
                 assert!(offset + block_count as u64 <= end);
                 let block = Block::new(offset, bs.trailing_zeros());
-                let buffer = Buffer::new(block_count * bs);
 
                 // The read will first try to establish a connection to
                 // the remote side before returning something we can await
                 // on. If that initial connection fails, we can retry.
-                let mut retry_needed = true;
                 let mut retry_count = 0;
-                while retry_needed {
-                    match read_only_parent.read(block, buffer.clone()).await {
-                        Ok(_) => {
-                            retry_needed = false;
+
+                let buffer = loop {
+                    let buffer = Buffer::new(block_count * bs);
```

Comment:
I don't see how to make this work, because `Buffer` isn't returned for the `Err` case. The only way I can see this working is if `read` had the signature:

```rust
    async fn read(
        &self,
        offset: Block,
        mut data: Buffer,
    ) -> Result<Buffer, (Buffer, CrucibleError)>;
```

which maybe it should? Just because `read` returned `Err` doesn't mean it should eat the Buffer...

Reply:
Alternatively,

```rust
    async fn read(
        &self,
        offset: Block,
        mut data: Buffer,
    ) -> (Buffer, Result<(), CrucibleError>);
```

Reply:
It moved out in 5b060cf